### PR TITLE
Binary Python wheels and sdist generated by Github Action

### LINKED ISSUE #1326: Have Python wheel
Having a Python wheel would speed up installation and remove the requirement for users to have a compilation pipeline set up.

Currently, if you pip install arbor, you will see output like:
```
Building wheels for collected packages: arbor
  Building wheel for arbor (setup.py) ... done
  Created wheel for arbor: filename=arbor-0.5-cp38-cp38-linux_x86_64.whl size=1322418 sha256=83fc9fe7d03f7afc73f69210bb9ea4373b08ee5d4c66180b9cfbb52c1aad35ac
  Stored in directory: /p/home/jusers/huisman1/jusuf/.cache/pip/wheels/09/18/68/ceac2887642e9c2e49a20630e3e5a94572bec0a994058e82dc
Successfully built arbor
```
If you take that wheel, it's perfectly installable in other envs/machines right now (just tested), as long as you match the platform. Turning that into manylinux (most recent version is 2014, see [PEP 599](https://www.python.org/dev/peps/pep-0599/)) seems like it shouldn't be so difficult (famous last words, I know).

### PR DESCRIPTION
**Pull request by brenthuisman**
## Changes

* Auto-generate binary wheels, triggered by git tags 'v*'.
  * Wheels are generated for Python versions 3.6-3.9, for Linux using the [PyPA manylinux-2014 image](https://github.com/pypa/manylinux) and MacOS using macos-latest (10.15 at time of writing).
  * The Action generates `sdist` for any other platforms.
  * The cibuildwheel action automatically prepares the wheels using [auditwheel](https://cibuildwheel.readthedocs.io/en/stable/options/#repair-wheel-command), which means bundling of external deps (e.g. libxml2 for NeuroML).
  * The output is all that's needed for publication of a new version to (Test)PyPI.
* `setup.py` defaults `neuroml` to on.
* A small change to `CMakeLists.txt` was required to build the wheels: based on this [Pybind hint](https://pybind11.readthedocs.io/en/stable/compiling.html#findpython-mode) `Development.Module`is now required, and not `Development` when CMake searches for Python. I confirm this change is needed to make Arbor build in the manylinux docker images, but I do not understand what the change implies, other than that we now don't link to `libpython`.
* (Unrelated to the above) Only run the `ebrains.yml` action on commits to master.

## Not changed

* Vectorization is off
  * Arbor's vectorization requires AVX (on x86-64), but that might be too much of a demand on all users (Intel CPUs >2010, AMD CPUs >2013 come with AVX), so I've left it `off` right now.
  * Reference for archs: https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html
    * Building Arbor with vectorization on but an arch without AVX support fails.

## Example

Example output (scroll down): https://github.com/brenthuisman/arbor/actions/runs/614579564
The artifacts can be download from the Github Action run, and you'll be just a `twine` away from uploading a new version.  

Fixes #1326

### CODE REVIEW

**Comment by Helveg** [comment]
Lovin the commit messages, CI fidgeting, yay. On the topic of the `arbor` vs `testarbor` I'd say use the `dev` version tag. PyPI's semver allows you to give a version like `0.5dev123` for exactly this purpose.

**Comment by brenthuisman** [comment]
I think we're already doing that (see `VERSION`). However, @bcumming has been doing all PyPI testing and releasing manually, so this represents a change of procedure. Also, so far `VERSION` is updated by hand.

File: CMakeLists.txt [hunk 1]
```diff
@@ -246,7 +246,11 @@ endif()
 if(ARB_WITH_PYTHON)
     cmake_dependent_option(ARB_USE_BUNDLED_PYBIND11 "Use bundled pybind11" ON "ARB_WITH_PYTHON;ARB_USE_BUNDLED_LIBS" OFF)
 
-    find_package(Python3 ${arb_py_version} COMPONENTS Interpreter Development REQUIRED)
+    if(DEFINED ENV{CIBUILDWHEEL})
+        find_package(Python3 ${arb_py_version} COMPONENTS Interpreter Development.Module REQUIRED)
```
Comment (thorstenhater) [inline CMakeLists.txt:250]: This conditional seems weird, I would expect that we can always load either Development.Module, Development, or both and have success?

File: doc/install/python.rst [hunk 2]
```diff
@@ -73,7 +73,7 @@ be used to configure the installation:
 * ``--mpi``: Enable MPI support (requires MPI library).
 * ``--gpu``: Enable GPU support for NVIDIA GPUs with nvcc using ``cuda``, or with clang using ``cuda-clang`` (both require cudaruntime).
   Enable GPU support for AMD GPUs with hipcc using ``hip``. By default set to ``none``, which disables gpu support.
-* ``--vec``: Enable vectorization. This might require choosing an appropriate architecture using ``--arch``.
+* ``--vec``: Enable vectorization. This might require choosing an appropriate architecture using ``--arch``. Note that on x86-64 platforms compilation will fail if you enable vectorization, but the CPU or ``--arch`` does not support any form of AVX.
```
Comment (thorstenhater) [inline doc/install/python.rst:76]: ... but defaults to the detected 'native', see below.

File: setup.py [hunk 3]
```diff
@@ -163,8 +213,12 @@ def run(self):
     zip_safe=False,
     ext_modules=[cmake_extension('arbor')],
     cmdclass={
-        'build_ext': cmake_build,
-        'install':   install_command,
+        'build_ext':   cmake_build,
+        'install':     install_command,
+        'bdist_wheel': bdist_wheel_command,
+    } if WHEEL_INSTALLED else {
```
Comment (thorstenhater) [inline setup.py:219]: Maybe store `cmdclass` in a variable set in the first if/else above?

File: .github/workflows/ciwheel.yml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        uses: joerick/cibuildwheel@v1.9.0
+        with:
+          output-dir: dist
+        env:
+          MACOSX_DEPLOYMENT_TARGET: "10.15" #needed to undo some CIBW settings
+          CIBW_BEFORE_BUILD: python -m pip install numpy setuptools
+          CIBW_BUILD: "cp3?-macosx_x86_64"
+          CIBW_SKIP: "cp35-*"
+          CIBW_ARCHS_MACOS: x86_64 universal2
+          CIBW_BUILD_VERBOSITY: 3
```
Comment (thorstenhater) [inline .github/workflows/ciwheel.yml:47]: That's left from debugging?

File: .github/workflows/ciwheel.yml [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+          # CIBW_TEST_COMMAND: TODO
+
+      - name: Build wheels macos
+        if: ${{ startsWith(matrix.os, 'macos') }}
+        uses: joerick/cibuildwheel@v1.9.0
+        with:
+          output-dir: dist
+        env:
+          MACOSX_DEPLOYMENT_TARGET: "10.15" #needed to undo some CIBW settings
+          CIBW_BEFORE_BUILD: python -m pip install numpy setuptools
```
Comment (thorstenhater) [inline .github/workflows/ciwheel.yml:43]: This (and CIBW_SKIP) can be set once on the job level env

Comment (thorstenhater) [reply .github/workflows/ciwheel.yml:43]: In fact, it's probably cleaner to use not the cartesian product matrix, but instead the config variant to set these variables (see basic.yml for how this looks like). The macos specifics can be set always (or, if we are worried about pollution using `run ::set ENV=VAR`)

**Review by thorstenhater — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (brenthuisman) [reply CMakeLists.txt:250]: Unfortunately, local builds with `Development.Module` fail, and in-container builds fail with `Development` (this seems to be on purpose). The only piece of documentation I could find is what I linked, but I don't understand it, only that it works.

File: setup.py (hunk 3 above)
Comment (brenthuisman) [reply setup.py:219]: OK

File: .github/workflows/ciwheel.yml (hunk 4 above)
Comment (brenthuisman) [reply .github/workflows/ciwheel.yml:47]: Yep

File: .github/workflows/ciwheel.yml (hunk 5 above)
Comment (brenthuisman) [reply .github/workflows/ciwheel.yml:43]: The Action runs this step in a docker image in the case of Linux (the manylinux2014 image) that does not inherit those things. It's clearer that this step is 'special' by keeping it contained like this.

Come to think of it, I probably don't need to setup Python.

File: doc/install/python.rst (hunk 2 above)
Comment (brenthuisman) [reply doc/install/python.rst:76]: I know, but if you specify an arch that hasn't AVX, build will fail. That's a gotcha that I though worth mentioning.

File: setup.py (hunk 3 above)
Comment (brenthuisman) [reply setup.py:214]: Turns out this creates a problem: https://github.com/brenthuisman/arbor/runs/2031776731?check_suite_focus=true

File: .github/workflows/ciwheel.yml (hunk 5 above)
Comment (thorstenhater) [reply .github/workflows/ciwheel.yml:42]: Seems weird that the same keyword has different scope, but well.

**Review by thorstenhater — APPROVED** [review]
Looks good, issues have been clarified.

**Pull request merged**